qa/tasks/cephfs/test_volumes.py

```diff
@@ -4345,6 +4352,317 @@ def test_subvolume_user_metadata_list_and_remove_for_legacy_subvolume(self):
         # verify trash dir is clean.
         self._wait_for_trash_empty()
 
+
+class TestPausePurging(TestVolumesHelper):
+    '''
+    Tests related to config "mgr/volumes/pause_purging".
+    '''
+
+    CONF_OPT = 'mgr/volumes/pause_purging'
+
+    def tearDown(self):
+        # every test will change value of this config option as per its need.
+        # assure that this config option's default value is re-stored during
+        # tearDown() so that there's zero chance that it interferes with next
+        # test.
+        self.config_set('mgr', self.CONF_OPT, False)
+
+        # ensure purge threads have no jobs left from previous test so that
+        # next test doesn't have to face unnecessary complications.
+        self._wait_for_trash_empty()
+
+        super().tearDown()
+
+    def _get_sv_path(self, v, sv):
+        sv_path = self.get_ceph_cmd_stdout(f'fs subvolume getpath {v} {sv}')
+        sv_path = sv_path.strip()
+        # delete slash at the beginning of path
+        sv_path = sv_path[1:]
+
+        sv_path = os.path.join(self.mount_a.mountpoint, sv_path)
+        return sv_path
+
+    def _assert_sv_is_absent_in_trash(self, sv, sv_path, sv_files):
+        uuid = self.mount_a.get_shell_stdout('sudo ls volumes/_deleting').\
+            strip()
+
+        trash_sv_path = sv_path.replace('_nogroup', f'_deleting/{uuid}')
+        trash_sv_path = trash_sv_path.replace(sv, '')
+
+        try:
+            sv_files_new = self.mount_a.get_shell_stdout(
+                f'sudo ls {trash_sv_path}')
+        except CommandFailedError as cfe:
+            # in case dir for subvol including files in it are deleted
+            self.assertEqual(cfe.exitstatus, 2)
+            return
+
+        # in case dir for subvol is undeleted yet (but files inside it are).
+        for filename in sv_files:
+            self.assertNotIn(filename, sv_files_new)
+
+    def _assert_trashed_sv_is_unpurged(self, sv, sv_path, sv_files):
+        uuid = self.mount_a.get_shell_stdout('sudo ls volumes/_deleting').\
+                    strip()
+
+        trash_sv_path = sv_path.replace('_nogroup', f'_deleting/{uuid}')
+        trash_sv_path = trash_sv_path.replace(sv, '')
+        sv_files_new = self.mount_a.get_shell_stdout(f'sudo ls {trash_sv_path}').\
+            strip()
+
+        if len(sv_files) == len(sv_files_new):
+            return
+        else:
+            for filename in sv_files:
```

Comment:
This is reached when the number of files don't match. Why doesn't the test fail right away?

Reply:
Pointless to run the loop if number of files before and after aren't the same.

Reply:
I also mentioned the condition under which it is pointless in the same sentence.

Reply:
Sorry, you are confusing me @rishabh-d-dave 

First of all, there is return under the if condition and thereby the else block is not necessary. Secondly, if the number of files mismatch, the loop checks if the passed in file names are in the scanned list of files. I didn't really get what is being done and I went back in to the previous function (L#4400) which says:

> # in case dir for subvol is undeleted yet (but files inside it are).

So, if you had 10 files in `sv_files` and you get 8 files by scanning (i.e., in `sv_files_new`), then assert is surely going to be hit, isn't it?

Reply:
If the whole point of that loop is to fail the test, then there are better, non-cryptic ways to doing so.

Reply:
Since as per our chat on Slack the name of missing doesn't really help and since we don't want any optimization, I am replacing this entire by the for loop it contains.